[typescript-fetch] add samples of nullable enum

berlysia opened the pull request:
resolves #7745 

<!-- Please check the completed items below -->
### PR checklist
 
- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] Pull Request title clearly describes the work in the pull request and Pull Request description provides details about how to validate the work. Missing information here may result in delayed response from the community.
- [x] If contributing template-only or documentation-only changes which will change sample output, [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) beforehand.
- [x] Run the shell script `./bin/generate-samples.sh`to update all Petstore samples related to your fix. This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit as it would merge with master. These must match the expectations made by your contribution. You may regenerate an individual generator by passing the relevant config(s) as an argument to the script, for example `./bin/generate-samples.sh bin/configs/java*`. For Windows users, please run the script in [Git BASH](https://gitforwindows.org/).
- [x] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master`
- [x] Copy the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) to review the pull request if your PR is targeting a particular programming language.

----

@TiFu (2017/07) @taxpon (2017/07) @sebastianhaas (2017/07) @kenisteward (2017/07) @Vrolijkx (2017/09) @macjohnny (2018/01) @topce (2018/10) @akehir (2019/07) @petejohansonxo (2019/11) @amakhrov (2020/02)

macjohnny commented on modules/openapi-generator/src/test/resources/3_0/petstore-with-fake-endpoints-models-for-testing.yaml:
```diff
@@ -1119,6 +1119,42 @@ paths:
       responses:
         200:
           description: The instance started successfully
+  /fake/enum-request:
```
I think this file should be copied rather than modified to avoid changing other test data

berlysia replied:
It's better to make a separate file of minimal yaml?
I'm doing this because I saw [this description](https://github.com/OpenAPITools/openapi-generator/blob/master/CONTRIBUTING.md#testing): 
> For new test cases, please add to the Fake Petstore spec

macjohnny replied:
ah, ok, yeah so its fine as it is, thanks

macjohnny replied:
there is still an issue with this test, could you fix it?
https://cloud.drone.io/OpenAPITools/openapi-generator/12016/1/5

berlysia commented:
it seems that I forgot to commit samples which are affected...
I couldn't see anything, sorry

I've regenerate all samples by using `./bin/generate-samples.sh` in https://github.com/OpenAPITools/openapi-generator/pull/7754/commits/d35283eeef6f0a629eb73f52e08a78e96bfc5da0

macjohnny commented:
@berlysia there are still some samples that need to be updated.

berlysia commented:
https://github.com/OpenAPITools/openapi-generator/pull/7754#discussion_r507746832 is fixed, but samples have difference.
It seems some another difference between `./bin/generate-samples.sh` result and reported on CI.
@macjohnny Which is prefer, or another?

<details>
<summary>on my env (42 changed, 329 files created)</summary>

```
> git status
On branch typescript-fetch-improvements
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
        modified:   samples/client/petstore/csharp/OpenAPIClient/README.md
        modified:   samples/client/petstore/csharp/OpenAPIClient/docs/DefaultApi.md
        modified:   samples/client/petstore/csharp/OpenAPIClient/docs/NullableClass.md
        modified:   samples/client/petstore/csharp/OpenAPIClient/src/Org.OpenAPITools/Api/DefaultApi.cs
        modified:   samples/client/petstore/csharp/OpenAPIClient/src/Org.OpenAPITools/Model/NullableClass.cs
        modified:   samples/client/petstore/perl/README.md
        modified:   samples/client/petstore/perl/docs/DefaultApi.md
        modified:   samples/client/petstore/perl/docs/NullableClass.md
        modified:   samples/client/petstore/perl/lib/WWW/OpenAPIClient/DefaultApi.pm
        modified:   samples/client/petstore/perl/lib/WWW/OpenAPIClient/Object/NullableClass.pm
        modified:   samples/client/petstore/php/OpenAPIClient-php/README.md
        modified:   samples/client/petstore/php/OpenAPIClient-php/docs/Api/DefaultApi.md
        modified:   samples/client/petstore/php/OpenAPIClient-php/docs/Model/NullableClass.md
        modified:   samples/client/petstore/php/OpenAPIClient-php/lib/Api/DefaultApi.php
        modified:   samples/client/petstore/php/OpenAPIClient-php/lib/Model/NullableClass.php
        modified:   samples/client/petstore/ruby-faraday/README.md
        modified:   samples/client/petstore/ruby-faraday/docs/DefaultApi.md
        modified:   samples/client/petstore/ruby-faraday/docs/NullableClass.md
        modified:   samples/client/petstore/ruby-faraday/lib/petstore/api/default_api.rb
        modified:   samples/client/petstore/ruby-faraday/lib/petstore/models/nullable_class.rb
        modified:   samples/client/petstore/ruby/README.md
        modified:   samples/client/petstore/ruby/docs/DefaultApi.md
        modified:   samples/client/petstore/ruby/docs/NullableClass.md
        modified:   samples/client/petstore/ruby/lib/petstore/api/default_api.rb
        modified:   samples/client/petstore/ruby/lib/petstore/models/nullable_class.rb
        modified:   samples/client/petstore/scala-gatling/bin/gatling/org/openapitools/client/model/Order.scala
        modified:   samples/openapi3/client/petstore/python/README.md
        modified:   samples/openapi3/client/petstore/python/docs/DefaultApi.md
        modified:   samples/openapi3/client/petstore/python/docs/NullableClass.md
        modified:   samples/openapi3/client/petstore/python/petstore_api/api/default_api.py
        modified:   samples/openapi3/client/petstore/python/petstore_api/models/nullable_class.py
        modified:   samples/schema/petstore/mysql/Model/NullableClass.sql
        modified:   samples/schema/petstore/mysql/mysql_schema.sql
        modified:   samples/server/petstore/jaxrs-jersey/src/gen/java/org/openapitools/api/FakeApi.java
        modified:   samples/server/petstore/jaxrs-jersey/src/gen/java/org/openapitools/api/FakeApiService.java
        modified:   samples/server/petstore/jaxrs-jersey/src/gen/java/org/openapitools/model/NullableClass.java
        modified:   samples/server/petstore/jaxrs-jersey/src/main/java/org/openapitools/api/impl/FakeApiServiceImpl.java
        modified:   samples/server/petstore/php-laravel/lib/app/Http/Controllers/DefaultController.php
        modified:   samples/server/petstore/php-laravel/lib/app/Models/NullableClass.php
        modified:   samples/server/petstore/php-laravel/lib/routes/api.php
        modified:   samples/server/petstore/php-lumen/lib/app/Http/Controllers/DefaultApi.php
        modified:   samples/server/petstore/php-lumen/lib/routes/web.php

Untracked files:
  (use "git add <file>..." to include in what will be committed)
        samples/client/petstore/groovy/bin/
        samples/client/petstore/kotlin-gson/bin/
        samples/client/petstore/kotlin-jackson/bin/
        samples/client/petstore/kotlin-json-request-string/bin/
        samples/client/petstore/kotlin-jvm-okhttp4-coroutines/bin/
        samples/client/petstore/kotlin-moshi-codegen/bin/
        samples/client/petstore/kotlin-nonpublic/bin/
        samples/client/petstore/kotlin-nullable/bin/
        samples/client/petstore/kotlin-okhttp3/bin/
        samples/client/petstore/kotlin-retrofit2-rx3/bin/
        samples/client/petstore/kotlin-retrofit2/bin/
        samples/client/petstore/kotlin-string/bin/
        samples/client/petstore/kotlin-threetenbp/bin/
        samples/client/petstore/kotlin-uppercase-enum/bin/
        samples/client/petstore/kotlin/bin/main/
        samples/client/petstore/kotlin/bin/test/
        samples/client/petstore/scala-gatling/bin/gatling/org/openapitools/client/model/InlineObject.scala
        samples/client/petstore/scala-gatling/bin/gatling/org/openapitools/client/model/InlineObject1.scala

no changes added to commit (use "git add" and/or "git commit -a")
```
</details>

<details>
<summary>on CI</summary>

```
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

	modified:   samples/client/petstore/csharp/OpenAPIClient/README.md
	modified:   samples/client/petstore/csharp/OpenAPIClient/docs/DefaultApi.md
	modified:   samples/client/petstore/csharp/OpenAPIClient/docs/NullableClass.md
	modified:   samples/client/petstore/csharp/OpenAPIClient/src/Org.OpenAPITools/Api/DefaultApi.cs
	modified:   samples/client/petstore/csharp/OpenAPIClient/src/Org.OpenAPITools/Model/NullableClass.cs
	modified:   samples/client/petstore/perl/README.md
	modified:   samples/client/petstore/perl/docs/DefaultApi.md
	modified:   samples/client/petstore/perl/docs/NullableClass.md
	modified:   samples/client/petstore/perl/lib/WWW/OpenAPIClient/DefaultApi.pm
	modified:   samples/client/petstore/perl/lib/WWW/OpenAPIClient/Object/NullableClass.pm
	modified:   samples/client/petstore/php/OpenAPIClient-php/README.md
	modified:   samples/client/petstore/php/OpenAPIClient-php/docs/Api/DefaultApi.md
	modified:   samples/client/petstore/php/OpenAPIClient-php/docs/Model/NullableClass.md
	modified:   samples/client/petstore/php/OpenAPIClient-php/lib/Api/DefaultApi.php
	modified:   samples/client/petstore/php/OpenAPIClient-php/lib/Model/NullableClass.php
	modified:   samples/client/petstore/ruby-faraday/README.md
	modified:   samples/client/petstore/ruby-faraday/docs/DefaultApi.md
	modified:   samples/client/petstore/ruby-faraday/docs/NullableClass.md
	modified:   samples/client/petstore/ruby-faraday/lib/petstore/api/default_api.rb
	modified:   samples/client/petstore/ruby-faraday/lib/petstore/models/nullable_class.rb
	modified:   samples/client/petstore/ruby/README.md
	modified:   samples/client/petstore/ruby/docs/DefaultApi.md
	modified:   samples/client/petstore/ruby/docs/NullableClass.md
	modified:   samples/client/petstore/ruby/lib/petstore/api/default_api.rb
	modified:   samples/client/petstore/ruby/lib/petstore/models/nullable_class.rb
	modified:   samples/openapi3/client/petstore/python/README.md
	modified:   samples/openapi3/client/petstore/python/docs/DefaultApi.md
	modified:   samples/openapi3/client/petstore/python/docs/NullableClass.md
	modified:   samples/openapi3/client/petstore/python/petstore_api/api/default_api.py
	modified:   samples/openapi3/client/petstore/python/petstore_api/models/nullable_class.py
	modified:   samples/schema/petstore/mysql/Model/NullableClass.sql
	modified:   samples/schema/petstore/mysql/mysql_schema.sql
	modified:   samples/server/petstore/jaxrs-jersey/src/gen/java/org/openapitools/api/FakeApi.java
	modified:   samples/server/petstore/jaxrs-jersey/src/gen/java/org/openapitools/api/FakeApiService.java
	modified:   samples/server/petstore/jaxrs-jersey/src/gen/java/org/openapitools/model/NullableClass.java
	modified:   samples/server/petstore/jaxrs-jersey/src/main/java/org/openapitools/api/impl/FakeApiServiceImpl.java
	modified:   samples/server/petstore/php-laravel/lib/app/Http/Controllers/DefaultController.php
	modified:   samples/server/petstore/php-laravel/lib/app/Models/NullableClass.php
	modified:   samples/server/petstore/php-laravel/lib/routes/api.php
	modified:   samples/server/petstore/php-lumen/lib/app/Http/Controllers/DefaultApi.php
	modified:   samples/server/petstore/php-lumen/lib/routes/web.php
```
</details>

macjohnny commented:
@berlysia in order to make it easier to review this PR and avoid changing the samples of other generators, I would copy the https://github.com/OpenAPITools/openapi-generator/blob/ce4990ecc3ceba25a62d80e4880f91807a1b35b4/modules/openapi-generator/src/test/resources/3_0/petstore-with-fake-endpoints-models-for-testing.yaml  file instead of modifying it. @wing328 any objections?

berlysia commented:
I will add minimal yaml which contains nullable enum values. It's too much large spec to copy, I think.

berlysia commented on modules/openapi-generator/src/test/resources/3_0/typescript-fetch/enum.yaml:
```diff
@@ -0,0 +1,289 @@
+            $ref: "#/components/schemas/StringEnum"
+        - name: nullable-string-enum
+          in: query
+          schema:
+            allOf:
+              - $ref: "#/components/schemas/StringEnum"
+            type: string
+            nullable: true
```
Oops. https://github.com/OAI/OpenAPI-Specification/issues/1368#issuecomment-580375017
> In JSON Schema, allOf doesn't work like inheritance. It means that the value must conform independently to each of the listed subschemas, and it must also conform to any other conditions specified directly in the schema.

I'll remove this because quoted as above and I've found https://github.com/OAI/OpenAPI-Specification/issues/2244 . Generated codes are roughly usable but this yaml does not conform OpenAPI Spec, currently.

berlysia commented:
@macjohnny Please take another look, this PR just adds samples now.
This is why I'm doing this. Please point it out something misunderstood.

I've finally noticed that "referenced nullable enum" is one of the most hottest issue of OpenAPI Generator and OpenAPI Specification.

- (referenced enum) There is a difference between inlined and referenced object, especially enum.
  - https://github.com/OpenAPITools/openapi-generator/issues/2645
- (nullable reference) There is no well-defined notation for simple `null` value in OAS v3.0.
  - This will be unlocked in OAS v3.1, by introducing `type: null`.
  - https://github.com/OAI/OpenAPI-Specification/issues/1368
- (nullable enum) `enum` cannot accept `null` only by `nullable: true` , `null` must be described in `enums: [(here)]` with `type: ["string", "null"]` , in JSON Schema. But there is difference between `SomeEnum | null` and `enum SomeEnum { Foo = "foo", NULL = null }`.
  - https://github.com/OAI/OpenAPI-Specification/issues/1900#issuecomment-486772917

## How we define a nullable variation of defined schema

for OAS v3.0 (it's weird along with spec, but useful codes will be generated):
```
propertyName:
  nullable: true
  allOf:
    - $ref: "#/pointer/to/defined/schema"
```

for OAS v3.1(not works currently because it's not implemented):
```
propertyName:
  oneOf:
    - type: "null"
    - $ref: "#/pointer/to/defined/schema"
```

see. https://github.com/OAI/OpenAPI-Specification/issues/1368#issuecomment-688368328 and https://github.com/OAI/OpenAPI-Specification/issues/1368#issuecomment-688413661

macjohnny reviewed (approved):
LGTM
